**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Gael Isaac Jurado Romero
*   MATRÍCULA: A01841366


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [1]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [2]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [3]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [4]:
# runtime está en minutos: 5 horas = 300 minutos
query = sqla.text("""
  SELECT *
  FROM movie
  WHERE runtime > 300
""")
pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


**Respuesta 1.** Hay **5 películas** con duración mayor a 5 horas (`runtime` > 300 minutos). Cuatro son entradas tituladas *Film Series*, con duraciones muy altas (hasta 3,892 minutos ≈ 65 horas) que sugieren que agrupan varias películas de una misma saga, y una es *Navarasa: Limited Series* (312 minutos). La columna `available_globally` aparece como bytes porque PyMySQL devuelve así los campos `BIT(1)` (`b'\x00'` = 0 y `b'\x01'` = 1).

2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [5]:
# available_globally = 0 indica que el título NO está disponible fuera de EU (es decir, solo en EU)
query = sqla.text("""
  SELECT ROUND(100 * SUM(available_globally = 0) / COUNT(*), 2) AS porcentaje_solo_eu
  FROM movie
  WHERE available_globally IS NOT NULL
""")
pd.read_sql(query, conn)

,porcentaje_solo_eu
0,83.57


**Respuesta 2.** El **83.57 %** de las películas está disponible únicamente en EU. El campo `available_globally` indica si el título está disponible fuera de EU, por lo que `available_globally = 0` significa que solo está en EU. El porcentaje se calcula directamente en SQL, excluyendo los registros con `available_globally` en `NULL`.

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [6]:
query = sqla.text("""
  SELECT DISTINCT locale
  FROM movie
  ORDER BY locale
""")
pd.read_sql(query, conn)

,locale
0,None
1,en


In [7]:
query = sqla.text("""
  SELECT COUNT(*) AS locale_nulos
  FROM movie
  WHERE locale IS NULL
""")
locale_nulos = int(pd.read_sql(query, conn).loc[0, 'locale_nulos'])
locale_nulos

11255

**Respuesta 3.** Los idiomas o regiones originales registrados en `movie` son únicamente `en` (inglés) y `NULL` (sin dato; `None` en Python, que según la versión de pandas se muestra como `None` o `NaN`). Hay **11,255 registros con `locale` NULL**, de un total de 11,787 películas.

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [8]:
# La collation de title y original_title ignora mayúsculas y acentos; CAST ... AS BINARY compara los textos exactos
query = sqla.text("""
  SELECT COUNT(CASE WHEN title <> original_title THEN 1 END)                                 AS diferentes_con_collation,
         COUNT(CASE WHEN CAST(title AS BINARY) <> CAST(original_title AS BINARY) THEN 1 END) AS diferentes_exactos
  FROM movie
""")
diferentes = pd.read_sql(query, conn)
titulos_diferentes = int(diferentes.loc[0, 'diferentes_exactos'])
diferentes

,diferentes_con_collation,diferentes_exactos
0,3947,3964


In [9]:
# Se compara con el resultado de la pregunta 3 (registros con locale NULL)
pd.DataFrame({
    'locale_nulos': [locale_nulos],
    'titulos_diferentes': [titulos_diferentes],
    'coinciden': [locale_nulos == titulos_diferentes],
    'restantes': [locale_nulos - titulos_diferentes]
})

,locale_nulos,titulos_diferentes,coinciden,restantes
0,11255,3964,False,7291


In [10]:
# Características de los registros: ¿tienen locale y/o original_title nulos?
query = sqla.text("""
  SELECT locale IS NULL AS locale_nulo,
         original_title IS NULL AS original_title_nulo,
         COUNT(*) AS peliculas
  FROM movie
  GROUP BY locale_nulo, original_title_nulo
  ORDER BY locale_nulo DESC, original_title_nulo
""")
pd.read_sql(query, conn)

,locale_nulo,original_title_nulo,peliculas
0,1,0,3964
1,1,1,7291
2,0,1,532


In [11]:
# Ejemplos de películas del grupo restante (locale y original_title NULL)
query = sqla.text("""
  SELECT title, locale, original_title
  FROM movie
  WHERE locale IS NULL AND original_title IS NULL
    AND title IN ('Damsel', 'Lift', 'Hit Man')
  ORDER BY title
""")
pd.read_sql(query, conn)

,title,locale,original_title
0,Damsel,None,None
1,Hit Man,None,None
2,Lift,None,None


**Respuesta 4.**

* **Títulos diferentes.** 3,964 películas tienen un `original_title` distinto de `title` (comparación exacta). Con un `<>` simple el servidor devuelve 3,947, porque la collation de ambas columnas ignora mayúsculas y acentos; los 17 casos de diferencia son pares como *Veronica* / *Verónica*, que sí son títulos distintos.
* **¿Coinciden los resultados?** No. Hay 11,255 registros con `locale` NULL, pero solo 3,964 con títulos diferentes: faltan **7,291** registros por explicar. Todas las películas con título distinto tienen `locale` NULL, pero no todas las de `locale` NULL tienen un título distinto.
* **Características de los registros restantes.** Los 7,291 registros tienen `locale` NULL y también `original_title` NULL, es decir, no hay un título original registrado con el cual compararlos. Entre ellos hay películas en inglés (por ejemplo *Damsel*, *Lift* y *Hit Man*), por lo que no pueden considerarse de otro idioma. Además, las 532 películas con `locale = 'en'` tienen siempre `original_title` NULL.
* **Conclusión.** La suposición **no es válida**: un `locale` NULL no indica que la película esté en otro idioma, sino que el dato no está registrado. Solo en 3,964 de las 11,255 películas con `locale` NULL (≈ 35 %) hay evidencia de un título original distinto; en el resto (≈ 65 %) no hay información, y algunas son en inglés.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [12]:
# cumulative_weeks_in_top10 es un contador acumulado: su máximo por película es el total de semanas en el top 10
query = sqla.text("""
  SELECT m.title, MAX(vs.cumulative_weeks_in_top10) AS semanas_en_top10
  FROM movie AS m
  JOIN view_summary AS vs ON vs.movie_id = m.id
  GROUP BY m.id, m.title
  ORDER BY semanas_en_top10 DESC
  LIMIT 1
""")
pd.read_sql(query, conn)

,title,semanas_en_top10
0,The Boss Baby,22


**Respuesta 5.** La película que ha permanecido más tiempo en el top 10 es ***The Boss Baby***, con **22 semanas** acumuladas. Se consideraron todos los registros de `view_summary` (sin filtrar por período): como `cumulative_weeks_in_top10` es un contador acumulado, su valor máximo por película equivale al total de semanas que ha estado en el top 10.

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [13]:
# RANK() da la misma posición a los programas empatados, para no cortar un empate con LIMIT
query = sqla.text("""
  SELECT posicion, programa, temporadas
  FROM (
    SELECT tv.title AS programa,
           COUNT(*) AS temporadas,
           RANK() OVER (ORDER BY COUNT(*) DESC) AS posicion
    FROM tv_show AS tv
    JOIN season AS s ON s.tv_show_id = tv.id
    GROUP BY tv.id, tv.title
  ) AS ranking
  WHERE posicion <= 5
  ORDER BY posicion, programa
""")
pd.read_sql(query, conn)

,posicion,programa,temporadas
0,1,Grey's Anatomy,20
1,1,Naruto Shippuden,20
2,3,Heartland (2007),17
3,4,It's Always Sunny in Philadelphia,16
4,5,NCIS,15
5,5,Raw,15
6,5,Supernatural (2005),15


**Respuesta 6.** Los programas con más temporadas son *Grey's Anatomy* y *Naruto Shippuden* (20 temporadas cada uno, empatados en el 1.er lugar), *Heartland (2007)* (17), *It's Always Sunny in Philadelphia* (16) y, en el 5.º lugar, *NCIS*, *Raw* y *Supernatural (2005)* con 15 temporadas cada uno. Como hay un empate triple en la 5.ª posición, la consulta devuelve 7 filas en lugar de 5: un `LIMIT 5` habría dejado fuera, de forma arbitraria, a dos de los tres programas empatados.

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [14]:
query = sqla.text("""
  SELECT DISTINCT start_date, end_date
  FROM view_summary
  WHERE duration = 'SEMI_ANNUALLY'
  ORDER BY start_date
""")
intervalos = pd.read_sql(query, conn)
intervalos

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


**Respuesta 7.** La tabla `view_summary` tiene dos períodos semestrales (`duration = 'SEMI_ANNUALLY'`): del **2023-07-01 al 2023-12-31** (segundo semestre de 2023) y del **2024-01-01 al 2024-06-30** (primer semestre de 2024).

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [15]:
# Primer período semestral de 2024, tomado del resultado de la consulta 7
semestre = intervalos[pd.to_datetime(intervalos['start_date']).dt.year == 2024].iloc[0]
inicio, fin = semestre['start_date'], semestre['end_date']

# Las vistas están redondeadas, por lo que hay empates: se desempata por número de temporada
query = sqla.text("""
  SELECT s.season_number, s.title AS temporada, vs.views
  FROM tv_show AS tv
  JOIN season AS s ON s.tv_show_id = tv.id
  JOIN view_summary AS vs ON vs.season_id = s.id
  WHERE tv.title = :titulo
    AND vs.duration = 'SEMI_ANNUALLY'
    AND vs.start_date = :inicio
    AND vs.end_date = :fin
  ORDER BY vs.views DESC, s.season_number
""")
temporadas_grey = pd.read_sql(query, conn, params={'titulo': "Grey's Anatomy", 'inicio': inicio, 'fin': fin})
temporadas_grey

,season_number,temporada,views
0,1,Grey's Anatomy: Season 1,3600000
1,2,Grey's Anatomy: Season 2,3100000
2,3,Grey's Anatomy: Season 3,2900000
3,4,Grey's Anatomy: Season 4,2900000
4,5,Grey's Anatomy: Season 5,2900000
5,6,Grey's Anatomy: Season 6,2800000
6,7,Grey's Anatomy: Season 7,2700000
7,8,Grey's Anatomy: Season 8,2600000
8,9,Grey's Anatomy: Season 9,2500000
9,10,Grey's Anatomy: Season 10,2400000


**Respuesta 8.** El primer período semestral de 2024 es el intervalo del 2024-01-01 al 2024-06-30 obtenido en la consulta 7, y las 20 temporadas de *Grey's Anatomy* tienen vistas registradas en él. La temporada más vista es la **1** (3.6 millones de vistas), seguida de la 2 (3.1 millones) y de las temporadas 3, 4 y 5 (2.9 millones cada una).

**Interpretación de los resultados.**

* **Las vistas disminuyen conforme avanza la serie.** El orden por vistas sigue casi el mismo orden de las temporadas: de 3.6 millones en la temporada 1 se baja gradualmente hasta 1.7 millones en las temporadas 17 y 18, es decir, la primera temporada tiene más del doble de vistas que ellas. Esto sugiere que la audiencia tiende a empezar la serie por el principio y a abandonarla de forma paulatina: muchos espectadores ven las primeras temporadas, pero pocos llegan a las más recientes.
* **Excepciones.** La temporada 19 (2.0 millones) supera a las temporadas 15 a 18, y la temporada 20 es un caso atípico con apenas 0.1 millones de vistas, muy por debajo del resto. Por ser la temporada más reciente, una posible explicación es que estuvo disponible poco tiempo durante el período, pero esta tabla no permite confirmarlo.
* **Precisión de los datos.** Las vistas están redondeadas a cientos de miles, por lo que hay empates (temporadas 3, 4 y 5; 12, 13, 14 y 19; 17 y 18). En esos casos el orden se definió por número de temporada y no indica diferencias reales de audiencia.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [16]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [17]:
# WHERE -> filtros booleanos, JOIN -> merge, ORDER BY -> sort_values
grey = tv_show.loc[tv_show['title'] == "Grey's Anatomy", ['id']].rename(columns={'id': 'tv_show_id'})
semestre_2024 = view_summary.loc[
    (view_summary['duration'] == 'SEMI_ANNUALLY')
    & (view_summary['start_date'] == inicio)
    & (view_summary['end_date'] == fin),
    ['season_id', 'views']
]
temporadas_grey_pd = (
    grey.merge(season[['id', 'tv_show_id', 'season_number', 'title']], on='tv_show_id')
        .merge(semestre_2024, left_on='id', right_on='season_id')
        .rename(columns={'title': 'temporada'})
        .sort_values(['views', 'season_number'], ascending=[False, True])
        .astype({'season_number': int})
        .loc[:, ['season_number', 'temporada', 'views']]
        .reset_index(drop=True)
)
temporadas_grey_pd

,season_number,temporada,views
0,1,Grey's Anatomy: Season 1,3600000
1,2,Grey's Anatomy: Season 2,3100000
2,3,Grey's Anatomy: Season 3,2900000
3,4,Grey's Anatomy: Season 4,2900000
4,5,Grey's Anatomy: Season 5,2900000
5,6,Grey's Anatomy: Season 6,2800000
6,7,Grey's Anatomy: Season 7,2700000
7,8,Grey's Anatomy: Season 8,2600000
8,9,Grey's Anatomy: Season 9,2500000
9,10,Grey's Anatomy: Season 10,2400000


In [18]:
# Verificación: el resultado en Pandas es idéntico al de la consulta 8 en SQL
temporadas_grey_pd.equals(temporadas_grey)

True

**Respuesta 9.** Con Pandas se obtiene el mismo resultado que con SQL en la consulta 8: las 20 temporadas, con las mismas vistas y en el mismo orden (la verificación con `equals` devuelve `True`). Equivalencias utilizadas: `WHERE` → filtros booleanos con `loc`, `JOIN` → `merge` y `ORDER BY` → `sort_values`; `inicio` y `fin` son las fechas del primer semestre de 2024 obtenidas en la consulta 7.

`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [19]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [20]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [21]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Colab Notebooks/MNA/TC5053 - Ciencia y analítica de datos/Actividad3_BD_Manipulacion"
os.chdir(DIR)

In [22]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [23]:
# stream() recorre los documentos de la colección; el id (doc.id) no viene en to_dict(), por eso se agrega
documentos = db.collection('EMPLOYEE').stream()
empleados = pd.DataFrame([{'id': doc.id, **doc.to_dict()} for doc in documentos]).set_index('id')
empleados[['emp_fname', 'emp_lname', 'emp_hiredate']]

,emp_fname,emp_lname,emp_hiredate
id,,,
8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


**Respuesta 10.** La colección se lee con `db.collection('EMPLOYEE').stream()`, que devuelve un documento por empleado. Los campos de cada documento se obtienen con `to_dict()` y su identificador con `doc.id`; como el `id` no forma parte de los campos, se agrega manualmente a cada registro y se usa como índice del DataFrame.

In [24]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

*   Anthropic. (2026). *Claude Code (basado en Claude Sonnet 5.5)* [Modelo de lenguaje grande], utilizado para generación de código, depuración y redacción de los análisis. https://claude.com/claude-code

---